### PR TITLE
Migration of GAN related tests and math test to Catch2 Testing

### PR DESCRIPTION
**Pull request by Abilityguy**
Regarding #2523

### CODE REVIEW

**Review by shrit — COMMENTED** [review]
Everything seems to be fine, can you resolve the merge conflict please, Thanks. :+1:

**Review by zoq — APPROVED** [review]
Thanks for putting this together, no comments from my side.

**Review by birm — APPROVED** [review]
Thanks for taking the time to do this!

**Review by shrit — APPROVED** [review]

**Review by Yashwants19 — COMMENTED** [review]
Windows build are not passing properly.

File: src/mlpack/tests/dcgan_test.cpp [hunk 1]
```diff
@@ -20,8 +20,7 @@
 
 #include <ensmallen.hpp>
 
-#include <boost/test/unit_test.hpp>
-#include "test_tools.hpp"
+#include "catch.hpp"
 #include "serialization.hpp"
```
Comment (Yashwants19) [inline src/mlpack/tests/dcgan_test.cpp:24]:
```suggestion
#include "test_catch_tools.hpp"
#include "serialization_catch.hpp"
```

I guess you can resolve the windows issue by including proper headers here and for other files too.

**Comment by zoq** [comment]
@Yashwants19 you are right, let's fix the issue before we merge this in.

**Comment by Yashwants19** [comment]
This is looking good. Let me restart github-workflow again.

**Comment by Yashwants19** [comment]
Okay, lets merge this. Thanks for the contribution. 👍

**Pull request merged**